### PR TITLE
Fix WaitForInit. Fixes #20

### LINKED ISSUE #20: testserver: broken with latest build
ts.WaitForInit(db) blocks indefinitely on my system (Ubuntu 16.04). Is it possible that the latest cockroachdb build breaks the testserver package?

I added a manual break after 5 seconds of iterating and logged the error and stdout/stderr from the testserver;

    func (ts *TestServer) WaitForInit(db *sql.DB) {
		cur, max := 0, 500
		for {
			// We issue a query that fails both on connection errors and on the
			// system database not existing.
			if _, err := db.Query("SHOW DATABASES"); err == nil {
				return
			} else if cur > max {
				log.Println(ts.Stdout())
				log.Println(ts.Stderr())
				panic(err)
			}
			cur++
			time.Sleep(time.Millisecond * 10)
		}
	}

Output (cleaned up):

    2016/12/04 15:52:31 downloading https://s3.amazonaws.com/cockroach/cockroach/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 to /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706, this may take some time
    2016/12/04 15:52:44 Using automatically-downloaded binary: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706
    2016/12/04 15:52:44 process 15520 started: /var/tmp/cockroach.linux-amd64.9111f757330103a89ccc9ef735d9c37867d71706 start --logtostderr --insecure --port=0 --http-port=0 --socket=/tmp/cockroach-testserver200397412/.s.PGSQL.26257 --store=/tmp/cockroach-testserver200397412
    waiting for database to initialize
    2016/12/04 15:52:50 CockroachDB node starting at 2016-12-04 15:52:44.767370201 +0100 CET

    --- FAIL: TestRunServer (19.29s)
    panic: pq: database "information_schema" does not exist [recovered]
        panic: pq: database "information_schema" does not exist

Testserver Stdout:

    build:      9111f75 @ 2016/12/04 13:07:50 (go1.7.3)
    admin:      http://m47:45529
    sql:        postgresql://root@m47:40251?sslmode=disable
    socket:     /tmp/cockroach-testserver200397412/.s.PGSQL.26257
    logs:       /tmp/cockroach-testserver200397412/logs
    store[0]:   path=/tmp/cockroach-testserver200397412
    status:     initialized new cluster
    clusterID:  fcbd6d02-c65b-4930-be7f-6859769ab779
    nodeID:     1

Testserver Stderr:

    2016/12/04 15:52:50 I161204 15:52:44.622880 1 cli/start.go:318  CockroachDB 9111f75 (linux amd64, built 2016/12/04 13:07:50, go1.7.3)
    I161204 15:52:44.723164 1 cli/start.go:334  starting cockroach node
    I161204 15:52:44.723274 1 cli/start.go:336  using local environment variables: COCKROACH_MAX_OFFSET
    W161204 15:52:44.723330 1 server/server.go:154  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
    I161204 15:52:44.723455 1 gossip/gossip.go:248  [n?] initial resolvers: []
    W161204 15:52:44.723495 1 gossip/gossip.go:1124  [n?] no resolvers found; use --join to specify a connected node
    I161204 15:52:44.725427 1 storage/engine/rocksdb.go:326  opening rocksdb instance at "/tmp/cockroach-testserver200397412"
    I161204 15:52:44.730046 50 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
    I161204 15:52:44.759296 1 server/config.go:443  1 storage engine initialized
    I161204 15:52:44.759864 1 server/node.go:419  [n?] store [n0,s0] not bootstrapped
    I161204 15:52:44.761700 63 storage/replica_proposal.go:348  [s1,r1/1:/M{in-ax},@c420276900] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 411350h52m53.761170701s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-12-04 14:52:44.761649824 +0000 UTC]
    I161204 15:52:44.762300 1 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
    I161204 15:52:44.762384 1 server/node.go:348  [n?] **** cluster fcbd6d02-c65b-4930-be7f-6859769ab779 has been created
    I161204 15:52:44.762404 1 server/node.go:349  [n?] **** add additional nodes by specifying --join=m47:40251
    I161204 15:52:44.762671 1 base/node_id.go:62  [n1] NodeID set to 1
    I161204 15:52:44.763014 1 storage/store.go:1223  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
    I161204 15:52:44.763051 1 server/node.go:432  [n1] initialized store [n1,s1]: {Capacity:105555197952 Available:61341224960 RangeCount:1 LeaseCount:1}
    I161204 15:52:44.763078 1 server/node.go:317  [n1] node ID 1 initialized
    I161204 15:52:44.763156 1 gossip/gossip.go:290  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"m47:40251" > attrs:<> locality:<> 
    I161204 15:52:44.763341 1 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
    I161204 15:52:44.763500 1 server/node.go:562  [n1] connecting to gossip network to verify cluster ID...
    I161204 15:52:44.763535 1 server/node.go:582  [n1] node connected via gossip and verified as part of cluster "fcbd6d02-c65b-4930-be7f-6859769ab779"
    I161204 15:52:44.763582 1 server/node.go:367  [n1] node=1: started with [[]=/tmp/cockroach-testserver200397412] engine(s) and attributes []
    I161204 15:52:44.763635 1 sql/executor.go:291  [n1] creating distSQLPlanner with address {tcp m47:40251}
    I161204 15:52:44.765237 28 storage/split_queue.go:103  [n1,split,s1,r1/1:/M{in-ax},@c420277200] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
    I161204 15:52:44.766902 28 storage/replica_command.go:2369  [n1,split,s1,r1/1:/M{in-ax},@c420277200] initiating a split of this range at key /Table/11 [r2]
    I161204 15:52:44.767156 1 server/server.go:633  [n1] starting http server at m47:45529
    I161204 15:52:44.767177 1 server/server.go:634  [n1] starting grpc/postgres server at m47:40251
    I161204 15:52:44.767198 1 server/server.go:635  [n1] advertising CockroachDB node at m47:40251
    I161204 15:52:44.767261 1 server/server.go:637  [n1] starting postgres server at unix:/tmp/cockroach-testserver200397412/.s.PGSQL.26257
    I161204 15:52:44.768282 106 server/updates.go:158  [n1,updates check] No previous updates check time.
    E161204 15:52:44.775774 29 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.775884 28 storage/queue.go:598  [n1,split,s1,r1/1:/{Min-Table/11},@c420277200] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
    E161204 15:52:44.775933 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.776003 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.776052 28 storage/split_queue.go:103  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
    I161204 15:52:44.776098 28 storage/replica_command.go:2369  [n1,split,s1,r2/1:/{Table/11-Max},@c420262480] initiating a split of this range at key /Table/12 [r3]
    E161204 15:52:44.780195 29 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.780334 28 storage/queue.go:598  [n1,split,s1,r2/1:/Table/1{1-2},@c420262480] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
    E161204 15:52:44.780492 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.780576 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.780705 28 storage/split_queue.go:103  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] splitting at keys [/Table/13/0 /Table/14/0]
    I161204 15:52:44.780758 28 storage/replica_command.go:2369  [n1,split,s1,r3/1:/{Table/12-Max},@c420262d80] initiating a split of this range at key /Table/13 [r4]
    E161204 15:52:44.784909 29 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.784938 28 storage/queue.go:598  [n1,split,s1,r3/1:/Table/1{2-3},@c420262d80] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
    I161204 15:52:44.785024 28 storage/split_queue.go:103  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] splitting at keys [/Table/14/0]
    I161204 15:52:44.785064 28 storage/replica_command.go:2369  [n1,split,s1,r4/1:/{Table/13-Max},@c4201ea900] initiating a split of this range at key /Table/14 [r5]
    E161204 15:52:44.785120 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.785190 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.785246 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789105 29 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789255 29 storage/queue.go:609  [n1,replicate,s1,r5/1:/{Table/14-Max},@c42024b680] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789312 110 storage/queue.go:609  [n1,replicate,s1,r1/1:/{Min-Table/11},@c420277200] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789372 110 storage/queue.go:609  [n1,replicate,s1,r2/1:/Table/1{1-2},@c420262480] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789426 110 storage/queue.go:609  [n1,replicate,s1,r3/1:/Table/1{2-3},@c420262d80] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    E161204 15:52:44.789477 110 storage/queue.go:609  [n1,replicate,s1,r4/1:/Table/1{3-4},@c4201ea900] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
    I161204 15:52:44.811683 134 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:m47:40251} Attrs: Locality:} ClusterID:fcbd6d02-c65b-4930-be7f-6859769ab779 StartedAt:1480863164763543657}
    I161204 15:52:45.353809 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160428, details: https://www.cockroachlabs.com/docs/beta-20160428.html
    I161204 15:52:45.353825 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160505, details: https://www.cockroachlabs.com/docs/beta-20160505.html
    I161204 15:52:45.353833 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160512, details: https://www.cockroachlabs.com/docs/beta-20160512.html
    I161204 15:52:45.353843 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160519, details: https://www.cockroachlabs.com/docs/beta-20160519.html
    I161204 15:52:45.353851 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160526, details: https://www.cockroachlabs.com/docs/beta-20160526.html
    I161204 15:52:45.353860 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160602, details: https://www.cockroachlabs.com/docs/beta-20160602.html
    I161204 15:52:45.353868 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160609, details: https://www.cockroachlabs.com/docs/beta-20160609.html
    I161204 15:52:45.353876 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160616, details: https://www.cockroachlabs.com/docs/beta-20160616.html
    I161204 15:52:45.353885 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160714, details: https://www.cockroachlabs.com/docs/beta-20160714.html
    I161204 15:52:45.353893 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160728, details: https://www.cockroachlabs.com/docs/beta-20160728.html
    I161204 15:52:45.353901 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160908, details: https://www.cockroachlabs.com/docs/beta-20160908.html
    I161204 15:52:45.353909 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160915, details: https://www.cockroachlabs.com/docs/beta-20160915.html
    I161204 15:52:45.353921 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20160929, details: https://www.cockroachlabs.com/docs/beta-20160929.html
    I161204 15:52:45.353930 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161006, details: https://www.cockroachlabs.com/docs/beta-20161006.html
    I161204 15:52:45.353938 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161013, details: https://www.cockroachlabs.com/docs/beta-20161013.html
    I161204 15:52:45.353946 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161027, details: https://www.cockroachlabs.com/docs/beta-20161027.html
    I161204 15:52:45.353956 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161103, details: https://www.cockroachlabs.com/docs/beta-20161103.html
    I161204 15:52:45.353964 106 server/updates.go:213  [n1,updates check] A new version is available: beta-20161201, details: https://www.cockroachlabs.com/docs/beta-20161201.html

### PR DESCRIPTION
**Pull request by mberhault**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach-go/22)

### CODE REVIEW

**Comment by tamird** [comment]
Not sure about this fix. Why does the old version work on macOS but not on Linux?

**Comment by mberhault** [comment]
It's not a platform thing, I pulled `github.com/lib/pq` and tried on my mac, same problem.

**Comment by tamird** [comment]
I have to rescind my earlier "this works on mac" - I just tried again and I
see the same timeout. My local lib/pq is at fcb9ef5.

**Comment by mberhault** [comment]
hum. circleci is failing on timeout. I figured I'd drop circleci and add a teamcity build for this, but same thing.

**Comment by mberhault** [comment]
circleci is not pulling this change, so the timeout is on the master version.
unsure about teamcity, I'm adding debugging

**Comment by mberhault** [comment]
ok, nevermind, circleci is grabbing the change, but there's something odd:
`~/.go_workspace/src/github.com/cockroachdb/cockroach-go` (the GOPATH) is at master and is what is being used.
the actual PR code is in `~/cockroach-go/`. Even if I run manually form within `~/cockroach-go`, the test still pulls in the master version.

Is there something odd in the way we pull dependencies? This branch is the same as `examples-go`, we basically run:
```
go get -d -t ./...
go test -v -i ./...
go test -v ./...
```

Same thing in teamcity, I've added: `cat $GOPATH/..../testserver.go` and `cat testserver/testserver.go` go the build steps, and it does show differences. see https://teamcity.cockroachdb.com/viewLog.html?buildId=83306&buildTypeId=Cockroach_CockroachGo&tab=buildLog

Clearly I'm doing something wrong. @tamird, @jordanlewis, ideas?

**Comment by mberhault** [comment]
PTAL: fixed teamcity: https://teamcity.cockroachdb.com/viewLog.html?buildId=83492&tab=buildResultsDiv&buildTypeId=Cockroach_CockroachGo
I'll drop circleCI in a subsequent PR.

**Comment by tamird** [comment]
so...why doesn't SHOW DATABASES work?

**Comment by mberhault** [comment]
looks like we can't talk to ourselves when using sockets:
```
I161207 14:35:23.327250 23 vendor/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp: lookup vintner: no such host"; Reconnecting to {"vintner:50612" <nil>}
```

This used to work fine, didn't it? at the very least, the tests in `examples-go` that actually write/read data were passing

**Comment by mberhault** [comment]
so yeah, this won't help tests actually trying to use the testserver since the whole point is to have a working cockroach node. Any idea when this may have changed?

**Comment by tamird** [comment]
I would guess this has been broken since whenever we started trying to talk
to ourselves internally, since grpc doesn't support dialing unix sockets
without some special machinery which we do not use.

**Comment by mberhault** [comment]
I supposed we need to figure out if we want to try and get around that, or if we should drop sockets.
Honestly, I'm fine with doing away with them, we can use @boljen's freeport method for this. Other than that, I'm not sure we need sockets for a whole lot. @bdarnell, care to weigh in?

**Comment by tamird** [comment]
We shouldn't be connecting to ourselves, so that's almost certainly a bug.

**Comment by tamird** [comment]
@mberhault were you able to reproduce that connection problem outside of this package's tests? I can't see to reproduce it on current master in docker.

**Comment by mberhault** [comment]
actually, I'm not. when I leave the same testserver that is failing, but attempt to connect to it with `cockroach sql --url='postgres://root@:26257?host=%2Ftmp%2Fcockroach-testserver331441468'`, `SHOW DATABASES` returns successfully.
does libpq persit errors somewhere?

**Comment by mberhault** [comment]
apparently it is. if I re-open a db handle on failure, it passes on the second attempt:
```
2016/12/07 16:36:30 Opening postgres://root@:26257?host=%2Ftmp%2Fcockroach-testserver578416532
2016/12/07 16:36:30 Wait for init
2016/12/07 16:36:30 Error: pq: database "information_schema" does not exist, opening new db
2016/12/07 16:36:30 WaitForInit: <nil>
```

@mjibson: could it be that libpq is trying to use `information_schema` somewhere before the actual query and therefore fails with a persistent error? Since the process just started up, the system tables may not exist yet.

**Comment by tamird** [comment]
Nah, lib/pq doesn't do any queries on its own - it's pretty bare. My guess
is that something gets fouled up on the cockroach side during process
startup. Perhaps it's time to bring @nvanbenschoten in.

**Comment by nvanbenschoten** [comment]
I can't seem to reproduce so it's tough to get close to the issue, but @mjibson was correct in saying that change in error text is almost certainly a result of https://github.com/cockroachdb/cockroach/pull/10196. The error is a little surprising to me, but I guess it's a result of hitting a `Server` before its SQL `Executor` has been [initialized](https://github.com/cockroachdb/cockroach/blob/83c4b3a32b11ab62703a5c078534d879d80fe6c9/pkg/sql/executor.go#L316). I'd assume we'd see the same error if we tried to query the table directly in that initialization period. Perhaps we should reconsider [the order](https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L636) that we start `Server` components and perform as much setup as possible before exposing ourselves on a port.

**Comment by tamird** [comment]
> Perhaps we should reconsider the order that we start Server components and perform as much setup as possible before exposing ourselves on a port.

Aren't we already doing that? We don't pass connections through to pgwire until:
https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L647

File: testserver/testserver.go [hunk 1]
```diff
@@ -225,19 +228,16 @@ func (ts *TestServer) PGURL() *url.URL {
...
+func (ts *TestServer) WaitForInit(db *sql.DB) (err error) {
+	for i := 0; i < 50; i++ {
+		if _, err = db.Query("SELECT 1"); err == nil {
 			return
 		}
-		time.Sleep(time.Millisecond * 10)
+		log.Printf("WaitForInit: %v", err)
+		time.Sleep(time.Millisecond * 100)
 	}
+	return
```
Comment (mjibson) [inline testserver/testserver.go]: This should return a real error since the wait didn't succeed.

Comment (mberhault) [inline testserver/testserver.go]: it is a real error, if we broke out it means that `err != nil`

**Comment by tamird** [comment]
I don't think we should merge this. We have a problem in Cockroach that
this papers over.

**Comment by mberhault** [comment]
agreed, while the timeout doesn't hurt, using `select 1` clearly does not do what we want. The goal is to notice when the system tables are there and the cluster ready to serve, so this should go back to being `show databases`. Keeping open for now.

**Comment by mberhault** [comment]
switched back to `SHOW DATABASES` to have a repeatable broken test to help with https://github.com/cockroachdb/cockroach/issues/12314

Also added a flag to specify a custom cockroach binary. Helpful to bisect over various binaries.

**Comment by tamird** [comment]
LGTM, probably best to squash this and remove the "fixes #20".

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYskxvbVAjYAdq8BniD:-KYskxvbVAjYAdq8BniE:bvhi416) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L79)):*
> ```Go
> var (
> 	sqlURLRegexp = regexp.MustCompile("sql:\\s+(postgresql:.+)\n")
> 	customBinary string
> ```

why not `customBinary = flag.String(...)`?

---

*[testserver/testserver.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYsl5p2p8QZR56v03AN:-KYsl5p2p8QZR56v03AO:b-d7fjtw) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L147)):*
> ```Go
> 	}
> 
> 	err = ts.WaitForInit(db)
> ```

`if err := ....; err != nil {` plz

---

*[testserver/testserver.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslOc7rBCZGMgVzHAv:-KYslOc7rBCZGMgVzHAw:b-xp0bf8) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L243)):*
> ```Go
> 
> // WaitForInit retries until a connection is successfully established.
> func (ts *TestServer) WaitForInit(db *sql.DB) (err error) {
> ```

while you're here, can you please not use a named return? I believe that's mjibson's confusion below.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYslY1v2w2EOnaayVa5:bsmrfhq) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
> ```Go
> // WaitForInit retries until a connection is successfully established.
> func (ts *TestServer) WaitForInit(db *sql.DB) (err error) {
> 	for i := 0; i < 50; i++ {
> ```

FWIW you can use https://godoc.org/github.com/cockroachdb/cockroach/pkg/util#RetryForDuration here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-:-KYslkKt5GfIl3KqgTyd:bm7vzli)*

**Comment by mberhault** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYskxvbVAjYAdq8BniD:-KYsmcq-wbS4jyOX4QKN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not `customBinary = flag.String(...)`?
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYsl5p2p8QZR56v03AN:-KYsn8Yqv_wvtb8_6E3o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L147)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`if err := ....; err != nil {` plz
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslOc7rBCZGMgVzHAv:-KYsnG2ukaDrOCE3_8XI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

while you're here, can you please not use a named return? I believe that's mjibson's confusion below.
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYsnGahwhChcc0dIbNu:b-n8jnql) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

FWIW you can use https://godoc.org/github.com/cockroachdb/cockroach/pkg/util#RetryForDuration here
</blockquote></details>

I can't, the whole point of this package is to avoid dependencies on `cockroachdb/cockroach`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*

**Comment by mberhault** [comment]
Squashed. I'm keeping the original commit message though, I'm not merging this until this works again.

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYsntGz5v6O-o3BLF0i:bhyeweg) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I can't, the whole point of this package is to avoid dependencies on `cockroachdb/cockroach`
</blockquote></details>

It is? I thought the point was to not to require building cockroach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*

**Comment by mberhault** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYuDZ8wCQV4w4I91Lxp:b-pkft2b) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It is? I thought the point was to not to require building cockroach.
</blockquote></details>

well, both. We don't pull `cockroachdb/cockroach` in this entire repo, which keeps a nice separation. I would prefer to keep it that way, unless we have a really good reason not to. A simple retry loop doesn't seem quite like enough.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*

**Pull request merged**